### PR TITLE
Bypass htaccess using a ip whitelist

### PR DESCRIPTION
**Pull request by lunika**
## Purpose

When a user has a trusted IP we wan't to deactivate the htaccess
protection while still keeping it active for other users.

## Proposal

Bypass htaccess using a whitelist of ip addresses for the following apps:

- [x] `richie`
- [x] `marsha`
- [x] `edxec`
- [x] `edxapp`

### CODE REVIEW

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 [hunk 1]
```diff
@@ -6,10 +7,10 @@ server {
   listen {{ edxapp_nginx_cms_port }};
   server_name localhost;
 
-  {% if activate_http_basic_auth or edxapp_activate_http_basic_auth -%}
+{% if http_basic_auth_enabled and edxapp_nginx_cms_bypass_htaccess_ip_whitelist | length == 0 %}
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: Following the same pattern as for `http_basic_auth_enabled`, can you define a `bypass_htaccess` variable to improve readability?

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 [hunk 2]
```diff
@@ -6,10 +7,10 @@ server {
...
   auth_basic "{{ http_basic_auth_message }}";
   auth_basic_user_file {{ http_basic_auth_user_file }};
-  {% endif %}
+{% endif %}
```
Comment (jmaupetit) [inline apps/edxapp/templates/services/nginx/configs/cms.conf.j2:22]: The whole configuration is rather complex. I have the feeling that we can improve readability by nesting blocks as follow:

```jinja
{% if http_basic_auth_enabled %}
{% if bypass_htaccess %}
  location @basicauth {
    auth_basic "{{ http_basic_auth_message }}";
    auth_basic_user_file {{ http_basic_auth_user_file }};

    try_files $uri @proxy_to_cms_app;
  }
{% else %}
  auth_basic "{{ http_basic_auth_message }}";
  auth_basic_user_file {{ http_basic_auth_user_file }};
{% endif %}
{% endif %}
```

WDYT?

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 (hunk 1 above)
Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2]: :+1:

File: apps/edxapp/templates/services/nginx/configs/cms.conf.j2 (hunk 2 above)
Comment (lunika) [reply apps/edxapp/templates/services/nginx/configs/cms.conf.j2:22]: :+1:

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**